File: _longest_repeating_character_replacement.ts
function characterReplacement(s: string, k: number): number {

    let map = new Map();
    let highFreq = 0;
    let longest =0;

    let l =0;
    let r =0;

    while(r<s.length){
        let val = s.charAt(r);

        map.set(val, map.get(val)+1 || 1)

        highFreq = Math.max(highFreq, map.get(val))

        while((r-l+1)-highFreq>k){
            let leftVal = s.charAt(l)
            map.set(leftVal, map.get(leftVal)-1 || 0)
            l++;
        }

        longest = Math.max(longest, r-l+1)

        r++;

    }

    return longest;

};



//Good ref- https://leetcode.com/problems/longest-repeating-character-replacement/solutions/2032968/javascript-solution-with-explanation-and-code-comments/

File: Python/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:

        res = 0

        map = {'A':0, 'B':0, 'C':0, 'D':0, 'E':0}

        l = 0
        r= 0

        while(r<len(s)):
            ele = s[r]
            if(not map.get(ele)):
                map[ele] = 0
            map[ele] = map[ele]+1

            max_val = max(map.values())
            if((r-l+1)-max_val>k):
                rem = s[l]
                map[rem] = map[rem]-1
                l = l+1
            
            if(r-l+1 > res):
                res = r-l+1


            r = r +1
        
        return res
